File: Java Programs/Chapter 3/TopologicalSort/src/topologicalsort/TopologicalSort.java

package topologicalsort;

import java.util.*;
import java.util.Scanner;


public class TopologicalSort {

    public static void TopologicalSort(int AdjMat[][],int n)
	{
		int InDegree[]=new int[n],i,j,k,u,w;
                int status[]=new int[n];//0-unvisited,1-in queue,2-visited
		Queue<Integer> q=new LinkedList<Integer>();
		for(i=0;i<n;i++)
                {
			InDegree[i]=0;
                        status[i]=0;
                }
		for(i=0;i<n;i++)
		{
			for(j=0;j<n;j++)
				if(AdjMat[i][j]!=0)
					InDegree[j]++;	
		}
                System.out.println("Indegree : ");
                for(i=0;i<n;i++)
			System.out.print(i+" : "+InDegree[i]+", ");
		for(i=0;i<n;i++)
			if(InDegree[i]==0)
				q.add(i);
		
		while(!q.isEmpty())
		{
		u=q.remove();
                status[u]=2;
		System.out.print("\nTopological Sort : ");
                System.out.print((u+1)+"\n");
		for(w=0;w<n;w++)
		{
			if(AdjMat[u][w]>0 && status[w]!=2)
				InDegree[w]--;
			if(InDegree[w]==0 && status[w]==0)
                        {
                            System.out.println("In queue : "+(w+1));
                            status[w]=1;
				q.add(w);
                        }
		}	
                System.out.println("\n\nIndegree : ");
                for(i=0;i<n;i++)
			System.out.print(i+" : "+InDegree[i]+", ");
		}
	}
    
    public static void main(String[] args) {
        int n;
	int AdjMat[][];

        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the no. of Vertices :");
        n = sc.nextInt();
        AdjMat = new int[n][n];
        System.out.println("Enter the Adjacency Matrix:");
	for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
		AdjMat[i][j] = sc.nextInt();
	System.out.println("Adjacency Matrix ");
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			System.out.print(AdjMat[i][j] + "   ");
		}
			System.out.println("\n");
	}
		
	TopologicalSort(AdjMat,n);
		
    }
}

File: Python Programs/unit3/TopologicalSort.py
from __future__ import print_function


queue = []
InDegree = [0] * 10


def TopologicalSort(AdjMat, n):
    status = [0] * 10  # 0-unvisited,1-in queue,2-visited
    for i in range(0, n):
        InDegree[i] = 0
        status[i] = 0
    for i in range(0, n):
        for j in range(0, n):
            if AdjMat[i][j] != 0:
                InDegree[j] += 1
    print("Indegree : ")
    for i in range(0, n):
        print("%d : %d" % (i, InDegree[i]))
    for i in range(0, n):
        if InDegree[i] == 0:
            queue.append(i)
    while len(queue) != 0:
        u = queue.pop(0)
        status[u] = 2
        print("\nTopological Sort : ")
        print("%d\n" % (u + 1))
        for w in range(0, n):
            if AdjMat[u][w] > 0 and status[w] != 2:
                InDegree[w] -= 1
            if InDegree[w] == 0 and status[w] == 0:
                print("In queue : %d", (w + 1))
                status[w] = 1
                queue.append(w)
                print("\n\nIndegree : ")
                for i in range(0, n):
                    print("%d : %d" % (i, InDegree[i]))


AdjMat = []
n = int(input("Enter the no. of Vertices :"))
print("Enter the Adjacency Matrix:")
for i in range(0, n):
    # replace raw_input with input if Python3
    arr = [
        int(x) for x in raw_input(
            "Enter Adjacency of %d node with space: \n" %
            i).split(" ")]
    AdjMat.append(arr)

print("Adjacency Matrix ")
for i in range(0, n):
    print("")
    for j in range(0, n):
        print(AdjMat[i][j], end=" ")
print("")
TopologicalSort(AdjMat, n)
